ccl/multiregionccl: TestMrSystemDatabase failed

ccl/multiregionccl.TestMrSystemDatabase [failed](https://tanzanite.cluster.engflow.com/invocations/default/3515e9e0-dfb1-4cad-9602-2011ae20e570?testReportRun=1&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on master @ [d45a65e08d45383aade2bcffdcdbe72a0cc279b1](https://github.com/cockroachdb/cockroach/commits/d45a65e08d45383aade2bcffdcdbe72a0cc279b1):


```
=== RUN   TestMrSystemDatabase
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestMrSystemDatabase966567483
    test_log_scope.go:81: use -show-logs to present logs inline

pkg/ccl/multiregionccl/multiregion_system_table_test.go:62: (TestMrSystemDatabase)
	NOTICE: .Locality() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().Locality() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
    multiregion_system_table_test.go:370: -- test log scope end --
test logs left over in: outputs.zip/logTestMrSystemDatabase966567483
--- FAIL: TestMrSystemDatabase (121.07s)
=== RUN   TestMrSystemDatabase/Sqlliveness
    multiregion_system_table_test.go:110: error scanning '&{0xc019146d80 <nil>}': pq: unsupported comparison: bytes to crdb_internal_region
    --- FAIL: TestMrSystemDatabase/Sqlliveness (0.19s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMrSystemDatabase.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-37660

ccl/multiregionccl.TestMrSystemDatabase [failed](https://tanzanite.cluster.engflow.com/invocations/default/478688e9-5d3a-453d-bddd-ea88693e3c8c?testReportRun=3&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on master @ [911f1ce389342459592d89fe7b78bcba3a2d265a](https://github.com/cockroachdb/cockroach/commits/911f1ce389342459592d89fe7b78bcba3a2d265a):


```
=== RUN   TestMrSystemDatabase
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestMrSystemDatabase1733990973
    test_log_scope.go:81: use -show-logs to present logs inline

pkg/ccl/multiregionccl/multiregion_system_table_test.go:62: (TestMrSystemDatabase)
	NOTICE: .Locality() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().Locality() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
    multiregion_system_table_test.go:370: -- test log scope end --
test logs left over in: outputs.zip/logTestMrSystemDatabase1733990973
--- FAIL: TestMrSystemDatabase (109.24s)
=== RUN   TestMrSystemDatabase/Sqlliveness
    multiregion_system_table_test.go:110: error scanning '&{0xc01e9f0c60 <nil>}': pq: unsupported comparison: bytes to crdb_internal_region
    --- FAIL: TestMrSystemDatabase/Sqlliveness (0.25s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMrSystemDatabase.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

ccl/multiregionccl.TestMrSystemDatabase [failed](https://tanzanite.cluster.engflow.com/invocations/default/49748a90-19e4-4a7c-b824-c364b3ce3e59?testReportRun=3&testReportShard=3&testReportAttempt=1#targets-Ly9wa2cvY2NsL211bHRpcmVnaW9uY2NsOm11bHRpcmVnaW9uY2NsX3Rlc3Q=) on master @ [cead8be66c30a7377dacbf1c678911d2cb17194d](https://github.com/cockroachdb/cockroach/commits/cead8be66c30a7377dacbf1c678911d2cb17194d):


```
=== RUN   TestMrSystemDatabase
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestMrSystemDatabase290297466
    test_log_scope.go:81: use -show-logs to present logs inline

pkg/ccl/multiregionccl/multiregion_system_table_test.go:62: (TestMrSystemDatabase)
	NOTICE: .Locality() called via implicit interface ApplicationLayerInterface;
HINT: consider using .ApplicationLayer().Locality() instead.
TIP: consider replacing the test server initialization from:
    ts, ... := serverutils.StartServer(t, ...)
    defer ts.Stopper().Stop(...)
to:
    srv, ... := serverutils.StartServer(t, ...)
    defer srv.Stopper().Stop(...)
    ts := srv.ApplicationLayer()

See also: https://go.crdb.dev/p/testserver-and-cluster-virtualization
    multiregion_system_table_test.go:377: -- test log scope end --
test logs left over in: outputs.zip/logTestMrSystemDatabase290297466
--- FAIL: TestMrSystemDatabase (96.74s)
=== RUN   TestMrSystemDatabase/Sqlliveness
    multiregion_system_table_test.go:110: error scanning '&{0xc017ff3680 <nil>}': pq: unsupported comparison: bytes to crdb_internal_region
    --- FAIL: TestMrSystemDatabase/Sqlliveness (0.17s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=3</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMrSystemDatabase.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/sql/catalog/lease/count.go
+++ b/pkg/sql/catalog/lease/count.go
@@ -23,6 +23,8 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/systemschema"
 	"github.com/cockroachdb/cockroach/pkg/sql/enum"
 	"github.com/cockroachdb/cockroach/pkg/sql/isql"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/regionliveness"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
@@ -58,8 +60,13 @@ func CountLeases(
 		systemDBVersion.AtLeast(clusterversion.V24_1_SessionBasedLeasingUpgradeDescriptor.Version())
 	leasingMode := readSessionBasedLeasingMode(ctx, settings)
 	whereClauses := make([][]string, 2)
+	containsSystemDatabase := false
+	forceMultiRegionQuery := false
 	for _, t := range versions {
 		versionClause := ""
+		if t.ID == keys.SystemDatabaseID {
+			containsSystemDatabase = true
+		}
 		if !forAnyVersion {
 			versionClause = fmt.Sprintf("AND version = %d", t.Version)
 		}
@@ -118,12 +125,24 @@ func CountLeases(
 			err := txn.WithSyntheticDescriptors(descsToInject,
 				func() error {
 					var err error
-					if cachedDatabaseRegions != nil && cachedDatabaseRegions.IsMultiRegion() {
+					if (cachedDatabaseRegions != nil && cachedDatabaseRegions.IsMultiRegion()) ||
+						forceMultiRegionQuery {
 						// If we are injecting a raw leases descriptors, that will not have the enum
 						// type set, so convert the region to byte equivalent physical representation.
 						count, err = countLeasesByRegion(ctx, txn, prober, regionMap, cachedDatabaseRegions, len(descsToInject) > 0, at, whereClause)
 					} else {
 						count, err = countLeasesNonMultiRegion(ctx, txn, at, whereClause)
+						// It's possible that our cached database descriptor is stale relative to the
+						// visible lease table descriptor. Because the lease manager uses rangefeeds to
+						// keep track of new descriptor versions, when converting to a multi-region
+						// system database its possible for the lease table descriptor update to precede
+						// the system database descriptor update.
+						if err != nil &&
+							pgerror.GetPGCode(err) == pgcode.UndefinedFunction &&
+							containsSystemDatabase {
+							forceMultiRegionQuery = true
+							return txn.KV().GenerateForcedRetryableErr(ctx, "forcing retry once with MR columns")
+						}
 					}
 					return err
 				})

